File: 0846/solution.java
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

class Solution {
    public boolean isNStraightHand(int[] hand, int groupSize) {
        Arrays.sort(hand);
        Map<Integer, Integer> counts = new HashMap<>();
        for (int num : hand) {
            counts.put(num, counts.getOrDefault(num, 0) + 1);
        }
        for (int num : hand) {
            while (counts.get(num) != 0) {
                for (int i = 0; i < groupSize; i++) {
                    int curr = counts.getOrDefault(num + i, 0);
                    if (curr == 0) {
                        return false;
                    }
                    counts.put(num + i, curr - 1);
                }
            }
        }
        return true;
    }
}

// time O(nlogn)
// space O(n)

File: 0846/solution.ts
function isNStraightHand(hand: number[], groupSize: number): boolean {
  const counts = new Map<number, number>();
  for (const num of hand) {
    counts.set(num, (counts.get(num) || 0) + 1);
  }
  hand.sort((a, b) => a - b);
  for (let num of hand) {
    while (counts.get(num) != 0) {
      for (let i = 0; i < groupSize; i++) {
        const curr = counts.get(num + i) || 0;
        if (curr == 0) {
          return false;
        }
        counts.set(num + i, curr - 1);
      }
    }
  }
  return true;
}
